### PR TITLE
doc: `nearcore` migration guide to 1.0.0

### PR DESCRIPTION
**Pull request by dj8yfo**
Md file is description of changes within [borsh update in nearcore](https://github.com/near/nearcore/pull/9432)

### CODE REVIEW

File: migration_guides/v0.10.2_to_v1.0.0_nearcore.md [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+ 2  error[E0433]: failed to resolve: could not find `maybestd` in `borsh`
+    --> core/primitives/src/action/delegate.rs:121:50
+     |
+ 121 |         ) -> ::core::result::Result<Self, borsh::maybestd::io::Error> {
+     |                                                  ^^^^^^^^ could not find `maybestd` i
+ n `borsh`
+```
+
+As `near-primitives` crate is not supposed to be used in `no_std` context, we can
+replace import with `std::io`:
```
Comment (frol) [inline migration_guides/v0.10.2_to_v1.0.0_nearcore.md:104]: Let's focus only on new classes of "errors" and skip the ones that were covered earlier.

File: borsh/src/lib.rs [hunk 2]
```diff
@@ -81,10 +81,12 @@ pub mod ser;
 
 pub use de::BorshDeserialize;
 pub use de::{from_reader, from_slice};
+
```
Comment (dj8yfo) [inline borsh/src/lib.rs:84]: these blanks change is so that release-plz picks up the commit for changelog, as the changes are otherwise outside of `borsh` or `borsh-derive` crates

File: migration_guides/v0.10.2_to_v1.0.0_nearcore.md [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+    InvalidPeerId = 8,
+    InvalidHash = 9,
+    InvalidEdge = 10,
+    InvalidDistanceVector = 11,
+    Blacklisted = 14,
+}
+```
+
+We fix it with `#[borsh(use_discriminant=false)]` as it will result in `12u8` value for `Blacklisted` 
+variant tag, used in serialization, same as before:
```
Comment (frol) [inline migration_guides/v0.10.2_to_v1.0.0_nearcore.md:250]:
```suggestion
We fix it with `#[borsh(use_discriminant=false)]` to preserve the behaviour of borsh before 1.0 release which serialized `ReasonForBan::Blacklisted` as 12 instead of 14 (borsh 0.10 and older [ignored explicit discriminant values in enum definitions](https://github.com/near/borsh-rs/issues/137)):
```

File: migration_guides/v0.10.2_to_v1.0.0_nearcore.md [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
++#[borsh(use_discriminant=false)]
+ pub enum ReasonForBan {
+```
+
+### 8. errors, similar to previous ones, in `near-network` and `near-state-viewer` packages
+
+There was a bunch of `borsh::maybestd` imports, which got replaced by their direct from-`std` counterparts.
+
+
+### 9. at this point project compiles, but we have and issue in ci with `cargo-deny --all-features check bans`
```
Comment (frol) [inline migration_guides/v0.10.2_to_v1.0.0_nearcore.md:268]: I feel this is out of scope of borsh migration itself, right? I would only mention this in the PR to nearcore to simplify the review.

File: migration_guides/v0.10.2_to_v1.0.0_nearcore.md [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+```
+
+```bash
+--- STDERR:              near-primitives action::delegate::tests::test_delegate_action_deserialization ---
+thread 'action::delegate::tests::test_delegate_action_deserialization' panicked at 'assertion failed: `(left == right)`
+  left: `Err(InvalidData)`,
+ right: `Err(InvalidInput)`', core/primitives/src/action/delegate.rs:172:9
+ ```
+
+ The `ErrorKind` in error in `borsh` hash changed, so we apply the following diff:
```
Comment (frol) [inline migration_guides/v0.10.2_to_v1.0.0_nearcore.md:327]: What do you mean by "hash" here?

File: migration_guides/v0.10.2_to_v1.0.0_nearcore.md [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
++++ b/chain/network/src/types.rs
+@@ -55,2 +55,3 @@ pub struct KnownProducer {
+ #[derive(borsh::BorshSerialize, borsh::BorshDeserialize, Debug, Clone, PartialEq, Eq, Copy)]
++#[borsh(use_discriminant=false)]
+ pub enum ReasonForBan {
+```
+
+### 8. errors, similar to previous ones, in `near-network` and `near-state-viewer` packages
+
+There was a bunch of `borsh::maybestd` imports, which got replaced by their direct from-`std` counterparts.
```
Comment (frol) [inline migration_guides/v0.10.2_to_v1.0.0_nearcore.md:265]: I would merge it with #6 and move it to the end of the list.

**Review by frol — COMMENTED** [review]
Overall, it is all great, but I want to mention that when I suggested having a migration guide I meant to write one for those who will need to migrate their projects to borsh 1.0.0, so the value is not that we teach people how to migrate nearcore to use borsh 1.0.0, but rather use nearcore to capture all the possible issues on that way and document them.

File: borsh/src/lib.rs (hunk 2 above)
Comment (frol) [reply borsh/src/lib.rs:84]: I think we don't necessarily need it as a line on the CHANGELOG, but rather we need a very prominent comment with a reference to the guide in the release notes to the release.

File: migration_guides/v0.10.2_to_v1.0.0_nearcore.md (hunk 5 above)
Comment (dj8yfo) [reply migration_guides/v0.10.2_to_v1.0.0_nearcore.md:327]: it's a typo of `has` -> `hash`

**Comment by frol** [comment]
@dj8yfo I have just cut 1.0.0-alpha.3 release with `#[borsh(skip)]`, which I believe will break more code in nearcore. Could you take a look, and update the PR and this guide?

File: migration_guides/v0.10.2_to_v1.0.0_nearcore.md (hunk 1 above)
Comment (dj8yfo) [reply migration_guides/v0.10.2_to_v1.0.0_nearcore.md:104]: It's the same class as one of above, but i specifically mentioned it with a ready recipe, as there's not much else, besides `hashbrown` and `nostd_io` in `borsh::__private` that branches on `std` support.
People also have been using `borsh::maybestd` for things, that are only available in `std`, like `std::sync::Arc`.

```diff
+#[cfg(feature = "std")]
+use std::io;
+#[cfg(not(feature = "std"))]
+use borsh::nostd_io as io;
```
